### ISSUE TITLE
Issue with Hyperion DASD Reserve/Release Support

### ISSUE THREAD

**Issue by arfineman** [SDL-Hercules-390/hyperion#300]
Given Hercules emulates a single channel path, DASD path grouping is of mute concern.  However, when a device receives and successfully executes a Reserve (x'B4') CCW, any subsequent Sense Path Group ID (x'34') CCW, should show the device as reserved. Upon receiving and successfully executing a  Release (x'94') CCW, the device reserve flag should be removed. 

This is a simple implementation, that could be used as a platform to fully support device reserve/release and DASD sharing. The final outcome should be something similar to VM's virtual reserve/release support.  

Of course in real life the implementation is a little more complicated. In 370 mode an I/O directed to a reserved device, receives a busy. When the device becomes available, a device end interrupt is sent to the channel path that received the busy condition.  In XA, the I/O is simply queued by the subsystem and is executed when the device becomes available.

**Comment by Fish-Git** [comment]
Thank you for reporting this Aaron!

**QUESTION:** Do you (or does anyone else) have a simple test case (perhaps a stand-alone test case?) that we can use to reproduce the problem (i.e. prove that it's occurring), and then later, once it's been found & fixed, to prove the bug now no longer exists? (has been squashed)

Having such a test program would help us a lot, and while I'm sure one of us could probably eventually throw something together, if you (or someone else) already has one (or could quickly throw one together for us), it would certainly expedite things somewhat.

Thanks!

**Comment by arfineman** [comment]
I opened this issue on request of Peter Jansen. He mentioned he has issues with Reserve/Release and I hope he chimes in. If he doesn't have a test case, then I'll provide a sample program on VM/370 using Diag20.
The fact is a reserved device is accessible from all channel paths from the system that issued the reserve.  So this will not protect against multiple applications sharing the same dasd on the same system. There are other methods of doing that, without reserve/release. 
Unless a single unit of dasd can be shared between multiple instances of Hercules, there is no application value to this implementation.  
Best regards,

**Comment by Peter-J-Jansen** [comment]
Thanks for opening this issue Aaron ! 

The reserve/release problem I have is for RACF database sharing between z/VM and z/OS, when I try to implement the sharing between multiple Hercules instances _and_ VM's as well. An example would be a z/VM RACF, a 2nd level z/OS RACF, and another Hercules (LPAR) z/OS. The manuals state that a full-pack minidisk is required that the VM's can MDISK / LINK to. Whatever I then do in the SYSTEM CONFIG re. User_Volume_List or _RDEV, a Hercules "ds" command shows the RACF database disk permanently being reserved.  IPL-ing z/OS on a 2nd Hercules instance on another PC using the Hercules shared DASD support, does not succeed but results in end endless waiting for the shared RACF database DASD. IPL-ing in the reverse order even causes z/VM to ABEND IQM002.

It looks like the chapter on `Sharing DASD among Multiple Virtual Machines and Other Systems
Using Concurrent Virtual and Real Reserve/Release` on pg 697 in the [z/VM CP Planning and Administration Guide](https://www.vm.ibm.com/library/640pdfs/64617814.pdf) does not work on Hercules.

When I look at the SDL-Hyperion code (shared.c, ckddasd.c), it is clear that _no_ shared DASD CCW commands at all will be executed when the DASD is in the `reserved` state by another Hercules instance. But as the current implementation always just _waits_ until the DASD is being `released` (370 mode not checked), and as the `reserved` state is verry short nowadys, I now doubt that my problem is caused by the in principle incorrect Hercules implementation.

The solution to my problem was to give up on the idea that multiple VM's should be able to share the RACF database using virtual reserve/release, and thus just DEDICATE (or ATTACH) the RACF database to RACFVM. Then everything works perfectly. I could still also make z/OS RACF run under z/VM, by defining another Hercules shared DASD to localhost, and also DEDICATE / ATTACH that effectively shared RACF database with a different CCUU devnum.

In short, I do have a reserve/release problem, but I'm not sure my problem is caused by Hercules. But the problem as you opened it is definitely valid, even if my problem is not a good test case for it.

Thanks for all your efforts.

Cheers,

Peter

**Comment by arfineman** [comment]
I wrote a small test utility to test the eventual fix of this issue. From what I can see, Hercules support of reserve/release is limited to a few commands that modern operating systems use during initialization and termination for dasd grouping. Although the path group id is preserved, no flags are set to show the status of the device on subsequent queries. 
I think I can provide the necessary information to have this fixed and a program to test it.

From what I understand from the DASD sharing add-ons, they mostly operate on a user defined start/end protocol and I don't expect them to work in a real environment with VM, MVS or RACF.

In your case, I can guarantee that VM's virtual reserve release will work, as long as you run everything second level under a single VM system. 
That means you need to define your RACF data base as a full pack minidisk (0-END) with MWV as the mode. 
In this scenario the first level VM system will manage the reserve/release CCWs from the second level system and responds correctly to their queries. And it will also issue a real Reserve/Release, that Hercules just ignores.   
HTH

**Comment by Peter-J-Jansen** [comment]
Thanks for the reply Aaron, it certainly helps !


> I think I can provide the necessary information to have this fixed and a program to test it.

That would be excellent, and I'd like to test your fixes as well, to see if it perhaps resolved my problem.

> From what I understand from the DASD sharing add-ons, they mostly operate on a user defined start/end protocol and I don't expect them to work in a real environment with VM, MVS or RACF.

The `user defined start/end protocol` is actually the start/end of a complete DASD CCW program. So in case of a DASD being reserved, the _whole CCW program_ will wait until that DASD has been released. The problem that you identified is that certain DASD CCW's should _not_ wait, but perhaps this only affects performance. But indeed, if this design deficiency would only affect performance, what else is wrong causing my problem ? 

> In your case, I can guarantee that VM's virtual reserve release will work, as long as you run everything second level under a single VM system.

I have no doubt about that. In another scenario (without RACF), I run several (base) sysplex z/OS systems 2nd level under z/VM, relying of MW**V** virtual release/reserve, and it works perfectly well.

> That means you need to define your RACF data base as a full pack minidisk (0-END) with MWV as the mode.

I am doing exactly that, and also use these SYSTEM CONFIG statements :

```
RDEVICE rdev1 TYPE DASD SHARED YES
RDEVICE rdev2 TYPE DASD SHARED YES

USER_VOLUME_LIST volser1 volser2
``` 
> In this scenario the first level VM system will manage the reserve/release CCWs from the second level system and responds correctly to their queries. And it will also issue a real Reserve/Release, that Hercules just ignores.

That's what the manuals suggest.  But all I see on Hercules, also without the second instance using Hercules shared DASD on another PC started, is that VM issues two Reserve/Release pairs at VM startup, and when it gets to using the RACF database on the shared DASD, _it just issues Release CCW's without ever a matching Release_. The Hercules Display Subchannel (DS) command shows the RACF databases as permanently Reserved.

How can Hercules possibly cause the Release CCW to not happen ? We can't have a VM bug here, can we ? 

The configuration which I cannot get to work is similar to the picture on pg 697 in the z/VM CP Planning and Administration manual which I mentioned before :

![image](https://user-images.githubusercontent.com/7361912/81662617-1fc35680-943e-11ea-80b9-0ba623231a9b.png)

In that picture, _virtual_ reserve/release can only help in case the 2nd VM issues a reserve whilst the 1st VM has it already reserved. Only then can CP avoid doing an I/O to the shared DASD and virtualize that reserve, by just waiting until the 1st VM does a release. In all other cases, a reserve from any VM must result into a real reserve against the shared DASD.

I'm looking forward to your suggestions and test program. And thanks again for all your help Aaron !
 Cheers,

Peter

**Comment by arfineman** [comment]
Hi Peter,
I'm not Hercules literate and was not aware of the DS command. I tested the command and see that  Hercules does keep track of the Reserve/Release CCWs. But that is not sufficient. Operating systems don't look at Hercules control blocks. Instead they expect the information from CCWs. 
This could be an easy fix. The device's reserve status has to be reflected in byte 0 bits 2-3 of the Sense Path Group ID (x'34') CCW. Since there are no other channel paths to a device under Hercules, it should set both bits on if the device is reserved and off if not.
Ideally Hercules should also preserve the single/multi path bit from the Set Path Group ID and return this information in byte 0 bit 4. 
It would also be nice if DS command displayed the path group id and single/multi path status.
Best regards,

**Comment by Peter-J-Jansen** [comment]
Hi Aaron,

OK, I see. That would be easy to implement for such SNID (x'34') CCW's coming from the same CPU / channel path as the one that issued the reserve. To allow SNID's (and the 3 others you mentioned (SPID, UR and RSTA) from any connected CPU / channel path / Hercules instance whilst the DASD is reserved would not be so easy, as the waiting while reserved is for a complete CCW program (i.e. the start/end sequence you mentioned). The granularity of the waiting would have to happen on a CCW basis., i.e. for all CCW's except the above 4.

Would you think that this could be the reason z/VM does not issue the release in my problem case ? I.e. z/VM doing a reserve, and then a SNID, finding these PGID bits off, then thinking "no need to release here", and thus no release, and thus z/OS waiting indefinitely ? z/OS does the reserve & release beautifully, and so does z/VM, but only when the DASD is DEDICATED to RACF, not when MDISK - LINKed.

Cheers.

Peter

**Comment by arfineman** [comment]
Yes, zVM does not issue a release, if it does not find the device reserved. The same is true when a virtual machine detaches a device. It does a X'34' and will only issue a release if it finds the device reserved to itself. 
As far as implementation you are correct and it will be a lot more complicated in a multi cpu environment. That will involve if a device is reserved to allow I/O from path that has the device reserved and queue the ones from other CPUs, until the device is released. Of course except for CCWs that I previously mentioned. 
Best regards

**Comment by Peter-J-Jansen** [comment]
Hi Aaron,

Problem solved ! You were right, also on how simple the fix was : 
```diff
hercules@pjjs12:~$ diff -uZ --color ~/Qsync/Source/Originals/SDL-hyperion/ckddasd.c ~/Qsync/Source/Changes/SDL-hyperion/ckddasd.c 
--- /home/hercules/Qsync/Source/Originals/SDL-hyperion/ckddasd.c	2020-03-30 18:54:26.000000000 +0200
+++ /home/hercules/Qsync/Source/Changes/SDL-hyperion/ckddasd.c	2020-05-16 03:19:56.000000000 +0200
@@ -120,6 +120,7 @@
 #define SPG_PATHSTAT_RESV       0x40    /* ...reserved bit setting   */
 #define SPG_PATHSTAT_UNGROUPED  0x80    /* ...ungrouped              */
 #define SPG_PATHSTAT_GROUPED    0xC0    /* ...grouped                */
+#define SPG_PATHSTAT_RESERVED   0x30    /* ...to channel + pg members*/
 #define SPG_PARTSTAT            0x30    /* Partitioning status bits..*/
 #define SPG_PARTSTAT_IENABLED   0x00    /* ...implicitly enabled     */
 #define SPG_PARTSTAT_RESV       0x10    /* ...reserved bit setting   */
@@ -5890,6 +5891,12 @@
                 | SPG_PARTSTAT_IENABLED
                 | SPG_PATHMODE_SINGLE;
 
+        /* If reserved then bits 2 & 3 must be set */
+        if (dev->reserved)
+        {
+            iobuf[0] |= SPG_PATHSTAT_RESERVED;
+        }
+
         /* Bytes 1-11 contain the path group identifier */
         memcpy (iobuf+1, dev->pgid, 11);
 
hercules@pjjs12:~$
```
The Hercules DS command shows the DASD in question now as NOT reserved, and inspection using the t+ccuu command shows that SNID's (0x34) are now correctly followed by Release (0x94) on the z/VM side. z/OS can share the DASD perfectly well with z/OS using Hercules shared DASD.

I doubt it that I would have been able to fix this without your help Aaron ! Thanks again !

Cheers,

Peter

P.S.: I'll close this Issue as soon as I have commited the fix.

**Comment by Peter-J-Jansen** [comment]
Commit 9fd1dff fixes this problem.

With many thanks to Aaron for spotting this in the first place and offering the solution to the problem !

Cheers,

Peter
